client.go

```diff
@@ -0,0 +1,249 @@
+type ClientConfig struct {
+	// APIKey is the Honeycomb authentication token. If it is specified during
+	// libhoney initialization, it will be used as the default API key for all
+	// events. If absent, API key must be explicitly set on a builder or
+	// event. Find your team's API keys at https://ui.honeycomb.io/account
+	APIKey string
+
+	// Dataset is the name of the Honeycomb dataset to which to send these events.
+	// If it is specified during libhoney initialization, it will be used as the
+	// default dataset for all events. If absent, dataset must be explicitly set
+	// on a builder or event.
+	Dataset string
+
+	// SampleRate is the rate at which to sample this event. Default is 1,
+	// meaning no sampling. If you want to send one event out of every 250 times
+	// Send() is called, you would specify 250 here.
+	SampleRate uint
+
+	// APIHost is the hostname for the Honeycomb API server to which to send this
+	// event. default: https://api.honeycomb.io/
+	APIHost string
+
+	// Transmission allows you to override what happens to events after you call
+	// Send() on them. By default, events are asynchronously sent to the
+	// Honeycomb API. You can use the MockOutput included in this package in
+	// unit tests, or use the WriterOutput to write events to STDOUT or to a
+	// file when developing locally.
+	Transmission transmission.Sender
+
+	// Logger defaults to nil and the SDK is silent. If you supply a logger here
+	// (or set it to &DefaultLogger{}), some debugging output will be emitted.
+	// Intended for human consumption during development to understand what the
+	// SDK is doing and diagnose trouble emitting events.
+	Logger Logger
+}
+
+// NewClient creates a Client with defaults correctly set
+func NewClient(conf ClientConfig) (*Client, error) {
+	if conf.SampleRate == 0 {
+		conf.SampleRate = defaultSampleRate
+	}
+	if conf.APIHost == "" {
+		conf.APIHost = defaultAPIHost
+	}
+	if conf.Dataset == "" {
+		conf.Dataset = defaultDataset
+	}
+
+	c := &Client{
+		logger: conf.Logger,
+	}
+	c.ensureLogger()
+
+	if conf.Transmission == nil {
+		c.transmission = &transmission.Honeycomb{
+			MaxBatchSize:         DefaultMaxBatchSize,
+			BatchTimeout:         DefaultBatchTimeout,
+			MaxConcurrentBatches: DefaultMaxConcurrentBatches,
+			PendingWorkCapacity:  DefaultPendingWorkCapacity,
+			UserAgentAddition:    UserAgentAddition,
+			Logger:               c.logger,
+			Metrics:              sd,
+		}
+	} else {
+		c.transmission = conf.Transmission
+	}
+	if err := c.transmission.Start(); err != nil {
+		c.logger.Printf("transmission client failed to start: %s", err.Error())
+		return nil, err
+	}
+
+	c.builder = &Builder{
+		WriteKey:   conf.APIKey,
+		Dataset:    conf.Dataset,
+		SampleRate: conf.SampleRate,
+		APIHost:    conf.APIHost,
+		dynFields:  make([]dynamicField, 0, 0),
+		fieldHolder: fieldHolder{
+			data: make(map[string]interface{}),
+		},
+		client: c,
+	}
+
+	return c, nil
+}
+
+func (c *Client) ensureTransmission() {
```

Comment:
All of these `ensure` methods are super not concurrency safe. Add a test which tries `Add()` etc on a shared, empty client from lots of goroutines and see what `go test -race` has to say about it. You're gonna need some locking if these fields are allowed to be set after `NewClient`.

However, for `logger` for example, you could instead add a `logf` method to `Client`, have everything call that instead of `c.logger.Printf`, and let that method handle the default case. You could include a `NullLogger` instance in Client for such cases even if you're not always using it.

`transmission` is harder but you could add methods for that too. Maybe. I'm not happy about the fact that if a method is called with a `nil` receiver, you end up creating and starting a new Transmission in each method call. It's such an edge case, maybe just blackhole the data instead... or don't handle the nil. A panic in a library is acceptable in the event of developer error.

Reply:
I sprinkled some `sync.Once` all over it. Tastes better, less filling.